target/scripts/build/packages.sh

```diff
@@ -0,0 +1,118 @@
+#! /bin/bash
+
+# -eE :: exit on error (do this in functions as well)
+# -u  :: show (and exit) when using unset variables
+# -o pipefail :: exit on error in pipes
+set -eE -u -o pipefail
+
+# shellcheck source=../helpers/log.sh
+source /usr/local/bin/helpers/log.sh
+
+QUIET='-qq' ; [[ ${LOG_LEVEL} =~ ^trace$ ]] && QUIET='-y'
+
+function _pre_installation_steps
+{
+  _log 'info' 'Starting package installation'
+  _log 'debug' 'Running pre-installation steps'
+
+  _log 'trace' 'Updating package signatures'
+  apt-get "${QUIET}" update
+
+  _log 'trace' 'Installing packages that are needed early'
+  apt-get "${QUIET}" install --no-install-recommends \
+    apt-transport-https apt-utils ca-certificates \
+    curl gnupg2 ssl-cert 2>/dev/null
+
+  _log 'trace' 'Upgrading packages'
+  apt-get "${QUIET}" upgrade
+
+  _log 'trace' 'Updating CA certificates'
+  update-ca-certificates
```

Comment:
>     * `apt-transport-https`, `ca-certificates`, `curl` were added later with the other packages. Is there a benefit to adding them here? (_no issue with that, just curious about the change_)

This is done in preparation for upcoming PRs that require these packages early (e.g. adding PPAs).

>     * `gnupg2`, `ssl-cert` are newly added via this PR. What are these needed for? `rspamd`? If so they should be deferred until that?

Indeed, `gnupg2` is for adding the PPA of `rspamd`. When I reviewed the installation log, I came across a few warnings from packages that want `ssl-cert` to be installed, so I put it here too.

I disagree that they should be deferred - this is more work for me without adding any benefit really, from my PoV. These packages are more or less essential and I think having them available early in the process is beneficial, installing `rspamd` or not.

>     * It doesn't look like we used `update-ca-certificates` previously, and while the `update` is done (_a little above the highlighted range_), you've also swapped `dist-upgrade` for `upgrade`?

The `update-ca-certificate` is done now because I saw some warnings in the logs about CA certificates, which are solved by updating (sorry, can't remember exactly what the message was, but `c_rehash` logged the warning).

> Is there a reason for preferring to switch from `dist-upgrade` to `upgrade`?

This was a lint from Hadolint about `dist-upgrade`, and the [AskUbuntu forum](https://askubuntu.com/a/226213) explains that `dist-upgrade` may remove packages. This is undesired, and Hadolint is right, `upgrade` is preferred I guess.

Reply:
> I disagree that they should be deferred - this is more work for me without adding any benefit really, from my PoV.

I understand that, and I know it's annoying to ask.

It's mostly for traceability in future. Many times if I attempt to wade through history searching for a commit message or PR that explains why something was added, even with packages it's not always evident beyond possible context of what the commit/PR is contributing.

In this case your comment at least adds some clarification if someone will search through the PR review process here. But it's not really related to `ci: enhance build process` and is very much belonging to your PR that will introduce `rspamd`.

Likewise, if someone goes through history for what was added to bring in `rspamd` support, they'd lack the context of these packages being associated to that, as the PR adding `rspamd` would presumably not even mention the relation.

---

Realistically... we've gained no new long-term maintainers since I got involved, and who knows when/if a new one will come along. I'm not going to block approval over this :+1: 

---

> Turns out, `update-ca-certificate` is not actually needed. I was unable to reproduce a warning message. I will remove it again.

I'm not against that btw, although I would assume it's part of the `upgrade` process when certs are updated, but I could be mistaken.

`c_rehash` was related to that I think. I tried to look back through `git blame` for where and why it was added, but the commit that introduced it was the "Final Migration" one with an invalid PR reference :laughing: 

---

> This was a lint from Hadolint about `dist-upgrade`, and the [AskUbuntu forum](https://askubuntu.com/a/226213) explains that `dist-upgrade` may remove packages. This is undesired, and Hadolint is right, `upgrade` is preferred I guess.

:+1:

Reply:
Alright, I can make an effort here :) I restored the original order in cde762061f9d08c54733583c57a94b8076bc8ef9 :)

Reply:
> Alright, I can make an effort here :)

I'm not sure what you're referring to with that?

I was requesting that `gnupg2` and `ssl-cert` be deferred to the relevant `rspamd` PR that would likely follow shortly after this is merged, or after the 11.2 release.

I just see them blended into the package list. Re-locating them wasn't of interest to me. I wanted them out of this PR, but did not want to mandate it. This is not a request to revert, I'll approve regardless of what you choose to do here, just a preference to have the packages belong in a PR they're intended for.

> I restored the original order in [cde7620](https://github.com/docker-mailserver/docker-mailserver/commit/cde762061f9d08c54733583c57a94b8076bc8ef9) :)

For `apt-transport-https`, `ca-certificates`, `curl` - I had said I did not mind these being re-located, I was just noting the observation as part of my review for tracking the shifting of packages in this PR :sweat_smile:

Reply:
Done in c6aa12436cbd0fc43f8dfac45d482f1e36689234